src/uu/tail/src/chunks.rs

```diff
@@ -81,3 +91,630 @@ impl<'a> Iterator for ReverseChunks<'a> {
+    /// Create a new [`LinesChunkBuffer`]
+    pub(crate) fn new(delimiter: u8, num_print: u64) -> Self {
+        Self {
+            delimiter,
+            num_print,
+            lines: 0,
+            chunks: VecDeque::new(),
+        }
+    }
+
+    /// Fills this buffer with chunks and consumes the reader completely. This method ensures that
+    /// there are exactly as many chunks as needed to match [`Self::num_print`] lines, so there are
+    /// in sum exactly `self.num_print` lines stored in all chunks. The method returns an iterator
+    /// over these chunks. If there are no chunks, for example because the piped stdin contained no
+    /// lines, or `num_print = 0` then `iterator.next` will return None.
+    ///
+    /// # Arguments
+    ///
+    /// * `reader`: A [`BufReader`] with an inner element implementing the [`Read`] trait.
+    ///
+    /// returns: Result<Iter<Box<LinesChunk, Global>>, Box<dyn UError, Global>>
+    pub(crate) fn fill(
+        &mut self,
+        reader: &mut BufReader<impl Read>,
+    ) -> UResult<Iter<Box<LinesChunk>>> {
+        let mut chunk = Box::new(LinesChunk::new(self.delimiter));
+
+        while (chunk.fill(reader)?).is_some() {
+            self.lines += chunk.lines as u64;
+            self.chunks.push_back(chunk);
+
+            let first = &self.chunks[0];
+            if self.lines - first.lines as u64 > self.num_print {
+                chunk = self.chunks.pop_front().unwrap();
+
+                self.lines -= chunk.lines as u64;
+            } else {
+                chunk = Box::new(LinesChunk::new(self.delimiter));
+            }
+        }
+
+        if !&self.chunks.is_empty() {
+            let length = &self.chunks.len();
+            let last = &mut self.chunks[length - 1];
+            if !last.get_buffer().ends_with(&[self.delimiter]) {
+                last.lines += 1;
+                self.lines += 1;
+            }
+        } else {
+            // chunks is empty when a file is empty so quitting early here
+            return Ok(self.chunks.iter());
+        }
+
+        // skip unnecessary chunks and save the first chunk which may hold some lines we have to
+        // print
+        let chunk = loop {
+            // it's safe to call unwrap here because there is at least one chunk and sorting out
+            // more chunks than exist shouldn't be possible.
+            let chunk = self.chunks.pop_front().unwrap();
+
+            // skip is true as long there are enough lines left in the other stored chunks.
+            let skip = self.lines - chunk.lines as u64 > self.num_print;
+            if skip {
+                self.lines -= chunk.lines as u64;
+            } else {
+                break chunk;
+            }
+        };
+
+        // calculate the number of lines to skip in the chunk
+        let skip_lines = if self.num_print >= self.lines {
+            0
+        } else {
+            (self.lines - self.num_print) as usize
+        };
```

Comment:
You could look into `saturating_sub` for this, something like:
```rust
self.lines.saturating_sub(self.num_print) as usize
```

Reply:
nice